### PR TITLE
feat: add AppID user agent parameter

### PR DESCRIPTION
**Pull request by yenfryherrerafeliz**
AppID is an user agent parameter that can be sourced from either client config, config file, and environment variable.

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/ClientResolver.php [hunk 1]
```diff
@@ -917,17 +926,41 @@ public static function _apply_http_handler($value, array &$args, HandlerList $li
         );
     }
 
+    public static function _apply_app_id($inputAppId, array &$args)
+    {
+        // AppId should not be longer than 50 chars
+        static $MAX_APP_ID_LENGTH = 50;
+        if (strlen($inputAppId) > $MAX_APP_ID_LENGTH) {
+            trigger_error("The provided or configured value for AppId, an user agent parameter, exceeds the maximum length of $MAX_APP_ID_LENGTH characters.", E_USER_WARNING);
```
Comment (stobrien89) [inline src/ClientResolver.php:934]: Let's try to make this multi-line

File: src/ClientResolver.php [hunk 2]
```diff
@@ -917,17 +926,41 @@ public static function _apply_http_handler($value, array &$args, HandlerList $li
         );
     }
 
+    public static function _apply_app_id($inputAppId, array &$args)
```
Comment (stobrien89) [inline src/ClientResolver.php:929]: let's stick with the convention we use in this class of calling the value `$value`

File: src/ClientResolver.php [hunk 3]
```diff
@@ -917,17 +926,41 @@ public static function _apply_http_handler($value, array &$args, HandlerList $li
...
+        }
+
+        $args['app_id'] = $inputAppId;
+    }
+
+    public static function _default_app_id(array $args)
+    {
+        return ConfigurationResolver::resolve(
+            'sdk_ua_app_id',
+            '',
```
Comment (stobrien89) [inline src/ClientResolver.php:944]: Not familiar with the spec— the default is an empty string?

File: tests/ClientResolverTest.php
Comment (stobrien89) [inline tests/ClientResolverTest.php:1]: We should add a test similar to some of the other user agent tests we have here: something like `testUserAgentAddsAppId`

**Review by stobrien89 — CHANGES_REQUESTED** [review]

File: src/ClientResolver.php (hunk 3 above)
Comment (yenfryherrerafeliz) [reply src/ClientResolver.php:944]: There is not a default value for app_id. If empty is returned here then, that means customer did not provided or configured an app_id value, and in that case this parameter is just ignored and not appended to the user agent header.

File: tests/ClientResolverTest.php
Comment (yenfryherrerafeliz) [reply tests/ClientResolverTest.php:1]: I think that is covered already [here](https://github.com/aws/aws-sdk-php/pull/2917/files#diff-e42a2aab192ff729219c1686f0383a82d6f38c2fb355bc7708a917213da71688R1579). The way I do the evaluation for the test is by checking the values from the user agent header from the request being sent.

**Review by stobrien89 — CHANGES_REQUESTED** [review]
Good to 🚢 after a changelog entry

**Review by stobrien89 — APPROVED** [review]

**Pull request merged**